File: java/Permutaciones.java
final class Permutaciones {
    static final int CATALOG_SIZE = 128;

    enum Op { R, L, U, D }
    enum PermOrder { ROWS_THEN_COLS, COLS_THEN_ROWS }
    enum PermRowsDir { LEFT, RIGHT }
    enum PermColsDir { UP, DOWN }

    record PermSpec(PermRowsDir rowsDir, PermColsDir colsDir, int nRows, int nCols, PermOrder order) {}

    static final Op[][] PAIRS = {
        {Op.R, Op.D}, {Op.R, Op.U}, {Op.L, Op.D}, {Op.L, Op.U},
        {Op.U, Op.R}, {Op.U, Op.L}, {Op.D, Op.R}, {Op.D, Op.L}
    };

    static PermSpec decode(int permIdx) {
        if (permIdx < 0 || permIdx >= CATALOG_SIZE) throw new IllegalArgumentException("permIdx invalido");
        int pairIdx = permIdx >> 4;
        int nRows = permIdx & 0x03;
        int nCols = (permIdx >> 2) & 0x03;
        Op[] pair = PAIRS[pairIdx];
        if (pair[0] == Op.R || pair[0] == Op.L) {
            return new PermSpec(
                pair[0] == Op.R ? PermRowsDir.RIGHT : PermRowsDir.LEFT,
                pair[1] == Op.D ? PermColsDir.DOWN : PermColsDir.UP,
                nRows, nCols, PermOrder.ROWS_THEN_COLS
            );
        }
        return new PermSpec(
            pair[1] == Op.R ? PermRowsDir.RIGHT : PermRowsDir.LEFT,
            pair[0] == Op.D ? PermColsDir.DOWN : PermColsDir.UP,
            nRows, nCols, PermOrder.COLS_THEN_ROWS
        );
    }

    static void applyU16(short[] data, int height, int width, int channels, int permIdx) {
        applySpecU16(data, height, width, channels, decode(permIdx));
    }

    static void applySpecU16(short[] data, int height, int width, int channels, PermSpec spec) {
        Object[] ops = specToOps(spec);
        applyOpU16(data, height, width, channels, (Op)ops[0], (Integer)ops[1]);
        applyOpU16(data, height, width, channels, (Op)ops[2], (Integer)ops[3]);
    }

    private static Object[] specToOps(PermSpec spec) {
        if (spec.order() == PermOrder.ROWS_THEN_COLS) {
            return new Object[] {
                spec.rowsDir() == PermRowsDir.RIGHT ? Op.R : Op.L, spec.nRows(),
                spec.colsDir() == PermColsDir.DOWN ? Op.D : Op.U, spec.nCols()
            };
        }
        return new Object[] {
            spec.colsDir() == PermColsDir.DOWN ? Op.D : Op.U, spec.nCols(),
            spec.rowsDir() == PermRowsDir.RIGHT ? Op.R : Op.L, spec.nRows()
        };
    }

    private static void applyOpU16(short[] data, int height, int width, int channels, Op op, int n) {
        short[] tmp = new short[data.length];
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                int srcI = i;
                int srcJ = j;
                int dstI = i;
                int dstJ = j;
                switch (op) {
                    case R -> srcJ = (j + i + n) % width;
                    case L -> dstJ = (j + i + n) % width;
                    case U -> srcI = (i + j + n) % height;
                    case D -> dstI = (i + j + n) % height;
                }
                int srcBase = (srcI * width + srcJ) * channels;
                int dstBase = (dstI * width + dstJ) * channels;
                System.arraycopy(data, srcBase, tmp, dstBase, channels);
            }
        }
        System.arraycopy(tmp, 0, data, 0, data.length);
    }
}

File: cs/Permutaciones.cs
using System;

namespace TTCrypto;

public static class Permutations
{
    public const int CatalogSize = 128;

    private enum Op { R, L, U, D }

    public readonly record struct PermSpec(PermRowsDir RowsDir, PermColsDir ColsDir, byte NRows, byte NCols, PermOrder Order);
    public enum PermOrder { RowsThenCols = 0, ColsThenRows = 1 }
    public enum PermRowsDir { Left = 0, Right = 1 }
    public enum PermColsDir { Up = 0, Down = 1 }

    private static readonly (Op Op1, Op Op2)[] Pairs =
    {
        (Op.R, Op.D), (Op.R, Op.U), (Op.L, Op.D), (Op.L, Op.U),
        (Op.U, Op.R), (Op.U, Op.L), (Op.D, Op.R), (Op.D, Op.L)
    };

    public static PermSpec Decode(byte permIdx)
    {
        if (permIdx >= CatalogSize) throw new ArgumentOutOfRangeException(nameof(permIdx));
        var pair = Pairs[permIdx >> 4];
        var nRows = (byte)(permIdx & 0x03);
        var nCols = (byte)((permIdx >> 2) & 0x03);
        return pair.Op1 switch
        {
            Op.R or Op.L => new PermSpec(
                pair.Op1 == Op.R ? PermRowsDir.Right : PermRowsDir.Left,
                pair.Op2 == Op.D ? PermColsDir.Down : PermColsDir.Up,
                nRows, nCols, PermOrder.RowsThenCols),
            _ => new PermSpec(
                pair.Op2 == Op.R ? PermRowsDir.Right : PermRowsDir.Left,
                pair.Op1 == Op.D ? PermColsDir.Down : PermColsDir.Up,
                nRows, nCols, PermOrder.ColsThenRows)
        };
    }

    public static void ApplyU16(ushort[] data, uint height, uint width, ushort channels, byte permIdx)
    {
        ApplySpecU16(data, height, width, channels, Decode(permIdx));
    }

    public static void ApplySpecU16(ushort[] data, uint height, uint width, ushort channels, PermSpec spec)
    {
        var (op1, n1, op2, n2) = SpecToOps(spec);
        ApplyOpU16(data, height, width, channels, op1, n1);
        ApplyOpU16(data, height, width, channels, op2, n2);
    }

    private static (Op op1, byte n1, Op op2, byte n2) SpecToOps(PermSpec spec)
    {
        if (spec.Order == PermOrder.RowsThenCols)
        {
            return (spec.RowsDir == PermRowsDir.Right ? Op.R : Op.L, spec.NRows,
                    spec.ColsDir == PermColsDir.Down ? Op.D : Op.U, spec.NCols);
        }
        return (spec.ColsDir == PermColsDir.Down ? Op.D : Op.U, spec.NCols,
                spec.RowsDir == PermRowsDir.Right ? Op.R : Op.L, spec.NRows);
    }

    private static void ApplyOpU16(ushort[] data, uint height, uint width, ushort channels, Op op, byte n)
    {
        var tmp = new ushort[data.Length];
        for (uint i = 0; i < height; i++)
        {
            for (uint j = 0; j < width; j++)
            {
                uint srcI = i, srcJ = j, dstI = i, dstJ = j;
                switch (op)
                {
                    case Op.R: srcJ = (uint)((j + i + n) % width); break;
                    case Op.L: dstJ = (uint)((j + i + n) % width); break;
                    case Op.U: srcI = (uint)((i + j + n) % height); break;
                    case Op.D: dstI = (uint)((i + j + n) % height); break;
                }
                var srcBase = ((int)srcI * (int)width + (int)srcJ) * channels;
                var dstBase = ((int)dstI * (int)width + (int)dstJ) * channels;
                Array.Copy(data, srcBase, tmp, dstBase, channels);
            }
        }
        Array.Copy(tmp, data, data.Length);
    }
}
